### PR TITLE
ui tests sandbox

### PR DESCRIPTION
**Pull request by nizienko**
added tasks `prepareUiTestingSandbox` and `runIdeForUiTests`

### CODE REVIEW

File: src/main/groovy/org/jetbrains/intellij/Utils.groovy [hunk 1]
```diff
@@ -85,16 +85,31 @@ class Utils {
...
+        "$sandboxDirectoryPath/system$suffix"
+    }
+
+    static def pluginsDir(@NotNull String sandboxDirectoryPath, String suffix) {
+        "$sandboxDirectoryPath/plugins$suffix"
+    }
+
+    @Deprecated
     static def configDir(@NotNull String sandboxDirectoryPath, boolean inTests) {
         def suffix = inTests ? "-test" : ""
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/Utils.groovy]: delegate to new methods

File: src/main/groovy/org/jetbrains/intellij/IntelliJPlugin.groovy [hunk 2]
```diff
@@ -306,28 +312,52 @@ class IntelliJPlugin implements Plugin<Project> {
...
+        configurePrepareSandboxTask(project, extension, PREPARE_SANDBOX_TASK_NAME, "")
+        configurePrepareSandboxTask(project, extension, PREPARE_TESTING_SANDBOX_TASK_NAME, "-test")
+        addRobotServerDependency(project, "0.9.1")
+        configurePrepareSandboxTask(project, extension, PREPARE_UI_TESTING_SANDBOX_TASK_NAME, "-uiTest")
+    }
+
+    private static void addRobotServerDependency(@NotNull Project project, @NotNull String robotServerVersion) {
+        Utils.info(project, "Configuring robot-server dependency")
+
+        project.repositories {
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/IntelliJPlugin.groovy]: it's better to delay it at least until the task will be actually run

File: src/main/groovy/org/jetbrains/intellij/IntelliJPlugin.groovy [hunk 3]
```diff
@@ -306,28 +312,52 @@ class IntelliJPlugin implements Plugin<Project> {
...
-            conventionMapping('destinationDir', { project.file(Utils.pluginsDir(extension.sandboxDirectory, inTest)) })
-            conventionMapping('configDirectory', { Utils.configDir(extension.sandboxDirectory, inTest).toString() })
+            conventionMapping('destinationDir', { project.file(Utils.pluginsDir(extension.sandboxDirectory, testSuffix)) })
+            conventionMapping('configDirectory', { Utils.configDir(extension.sandboxDirectory, testSuffix).toString() })
             conventionMapping('librariesToIgnore', { project.files(extension.ideaDependency.jarFiles) })
             conventionMapping('pluginDependencies', { extension.pluginDependencies })
             dependsOn(JavaPlugin.JAR_TASK_NAME)
+
+            if (taskName == PREPARE_UI_TESTING_SANDBOX_TASK_NAME) {
+                from project.zipTree(project.configurations.robotServerPluginImplementation.files.find { it.name.endsWith("zip") })
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/IntelliJPlugin.groovy]: there is a `singleFile` method, it might fit

Comment (zolotov) [reply src/main/groovy/org/jetbrains/intellij/IntelliJPlugin.groovy]: copying is also should be delayed to actually run. see `pluginDependencies` convention mapping, you should fill it with your plugin

File: src/main/groovy/org/jetbrains/intellij/tasks/DownloadRobotRegistryPluginTask.groovy [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import org.jetbrains.intellij.dependency.PluginDependency
+import org.jetbrains.intellij.dependency.PluginDependencyImpl
+
+class DownloadRobotRegistryPluginTask extends ConventionTask {
+    @Input
+    String robotServerVersion
+
+    @TaskAction
+    void download() {
+        def sourceUrl = "https://bintray.com/jetbrains/intellij-third-party-dependencies/download_file?file_path=org%2Fjetbrains%2Ftest%2Frobot-server-plugin%2F${robotServerVersion}%2Frobot-server-plugin-${robotServerVersion}.zip"
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/tasks/DownloadRobotRegistryPluginTask.groovy]: No need to download maven dependency manually. Gradle can do this for you. See instrument task, it downloads compiler from maven repo

File: src/main/groovy/org/jetbrains/intellij/tasks/DownloadRobotRegistryPluginTask.groovy [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        def outputDir = new File(project.buildDir, "robotServerPlugin")
+        outputDir.mkdirs()
+        ant.get(src: sourceUrl, dest: "$outputDir/robot-server-plugin.zip")
+        ant.unzip(src: "$outputDir/robot-server-plugin.zip", dest: outputDir)
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/tasks/DownloadRobotRegistryPluginTask.groovy]: Should be declared as output

File: src/main/groovy/org/jetbrains/intellij/IntelliJPlugin.groovy [hunk 6]
```diff
@@ -351,8 +346,9 @@ class IntelliJPlugin implements Plugin<Project> {
             dependsOn(JavaPlugin.JAR_TASK_NAME)
 
             if (taskName == PREPARE_UI_TESTING_SANDBOX_TASK_NAME) {
-                from project.zipTree(project.configurations.robotServerPluginImplementation.files.find { it.name.endsWith("zip") })
-                into Utils.pluginsDir(extension.sandboxDirectory, "-uiTest")
+                dependsOn(DOWNLOAD_ROBOT_SERVER_PLUGIN_TASK_NAME)
+                def downloadTask = project.tasks.findByName(DOWNLOAD_ROBOT_SERVER_PLUGIN_TASK_NAME) as DownloadRobotRegistryPluginTask
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/IntelliJPlugin.groovy]: You can do this on the call site, no need to have it in the common code

File: src/main/groovy/org/jetbrains/intellij/tasks/DownloadRobotRegistryPluginTask.groovy (hunk 4 above)
Comment (nizienko) [reply src/main/groovy/org/jetbrains/intellij/tasks/DownloadRobotRegistryPluginTask.groovy]: fixed

File: src/main/groovy/org/jetbrains/intellij/tasks/DownloadRobotRegistryPluginTask.groovy (hunk 5 above)
Comment (nizienko) [reply src/main/groovy/org/jetbrains/intellij/tasks/DownloadRobotRegistryPluginTask.groovy]: fixed

File: src/main/groovy/org/jetbrains/intellij/IntelliJPlugin.groovy (hunk 6 above)
Comment (nizienko) [reply src/main/groovy/org/jetbrains/intellij/IntelliJPlugin.groovy]: moved it to  `configureProjectAfterEvaluate` method

File: src/main/groovy/org/jetbrains/intellij/Utils.groovy [hunk 7]
```diff
@@ -85,19 +85,31 @@ class Utils {
         result
     }
 
+    static def configDir(@NotNull String sandboxDirectoryPath, String suffix) {
```
Comment (zolotov) [inline src/main/groovy/org/jetbrains/intellij/Utils.groovy]: I think all these methods might be inlined now

Comment (nizienko) [reply src/main/groovy/org/jetbrains/intellij/Utils.groovy]: they are invoked from two places. But I deleted old methods which is not using now

Comment (zolotov) [reply src/main/groovy/org/jetbrains/intellij/Utils.groovy]: But this is just concatenation, here is no logic anymore

Comment (nizienko) [reply src/main/groovy/org/jetbrains/intellij/Utils.groovy]: fixed

**Comment by nizienko** [comment]
> how often it's going to happen?

not often I hope, just decided to update it until the branch is not merged yet

**Pull request closed without merging**